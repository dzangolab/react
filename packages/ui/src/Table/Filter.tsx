import React, { useContext, useState } from "react";
import { DebounceInput } from "react-debounce-input";
import OutsideClickHandler from "react-outside-click-handler";

import { TableContext } from "./TableProvider";

import type { FilterProperties } from "./types";

function Filter({
  columnFilterValue,
  columnType,
  handleChange,
}: FilterProperties) {
  const { inputDebounceTime, filterIcons } = useContext(TableContext);

  const [expanded, setExpanded] = useState(false);

  const toggleExpand = () => {
    expanded ? setExpanded(false) : setExpanded(true);
  };

  if (columnType === "number") return null;

  return (
    <div
      className="filter-wrapper"
      onClick={(event_) => event_.stopPropagation()}
    >
      {
        <button onClick={toggleExpand}>
          {expanded ? (
            <img src={filterIcons?.expanded} />
          ) : (
            <img src={filterIcons?.notExpanded} />
          )}
        </button>
      }
      {expanded ? (
        <OutsideClickHandler
          onOutsideClick={() => {
            setExpanded(false);
          }}
        >
          <div role="dialog" className="filter-form-wrapper">
            <DebounceInput
              type="text"
              value={(columnFilterValue ?? "") as string}
              debounceTimeout={inputDebounceTime}
              onChange={(event_) => {
                handleChange(event_.target.value);
              }}
              placeholder={`Search...`}
              className=""
              size={10}
            />
          </div>
        </OutsideClickHandler>
      ) : null}
    </div>
  );
}

export default Filter;
