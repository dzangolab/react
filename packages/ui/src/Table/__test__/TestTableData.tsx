const data = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
];

const biggerData = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
  { id: 5, name: "tony" },
  { id: 6, name: "tony" },
  { id: 7, name: "tony" },
  { id: 8, name: "tony" },
  { id: 9, name: "tony" },
  { id: 10, name: "tony" },
  { id: 11, name: "tony" },
  { id: 12, name: "tony" },
  { id: 13, name: "tony" },
  { id: 14, name: "tony" },
  { id: 15, name: "tony" },
];

const fetcher = () => {
  return data;
};

const columns = [
  {
    accessorKey: "id",
    header: () => <span>ID</span>,
  },
  {
    accessorKey: "name",
    header: () => <span>Name</span>,
  },
];

export { columns, data, fetcher, biggerData };
