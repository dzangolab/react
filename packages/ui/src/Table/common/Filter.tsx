import React, { useContext, useState } from "react";
import { DebounceInput } from "react-debounce-input";
import OutsideClickHandler from "react-outside-click-handler";

import { TableContext } from "./TableContext";

import type { FilterProperties } from "./types";

const renderImage = (source?: string) => {
  return <img src={source} />;
};

function Filter({
  filterVariant = "text",
  /* eslint-disable-next-line unicorn/prevent-abbreviations */
  filterFn,
  placeholder,
  selectOptions,
  columnFilterValue,
  columnType,
  handleChange,
}: FilterProperties) {
  if (!filterVariant) return null;

  const { inputDebounceTime, filterIcons } = useContext(TableContext);

  const [expanded, setExpanded] = useState(false);

  const toggleExpand = () => {
    expanded ? setExpanded(false) : setExpanded(true);
  };

  return (
    <div onClick={(event_) => event_.stopPropagation()}>
      {
        <button onClick={toggleExpand}>
          {expanded
            ? renderImage(filterIcons?.expanded)
            : renderImage(filterIcons?.notExpanded)}
        </button>
      }
      {expanded ? (
        <OutsideClickHandler
          onOutsideClick={() => {
            setExpanded(false);
          }}
        >
          <div role="dialog" className="filter-form-wrapper">
            {filterVariant === "text" ? (
              <DebounceInput
                type="text"
                value={(columnFilterValue?.value ?? "") as string}
                debounceTimeout={inputDebounceTime}
                onChange={(event_) => {
                  handleChange({
                    value: event_.target.value,
                    filterFn: filterFn || "contains",
                  });
                }}
                placeholder={placeholder || `Search...`}
                className="filter-text-input"
                size={10}
              />
            ) : null}

            {filterVariant === "select" && selectOptions ? (
              <select
                value={(columnFilterValue?.value ?? "") as string}
                onChange={(event_) => {
                  handleChange({
                    value: event_.target.value,
                    filterFn: filterFn || "equals",
                  });
                }}
                className="filter-select-input"
              >
                {placeholder ? (
                  <option value="" disabled selected>
                    {placeholder}
                  </option>
                ) : null}

                {selectOptions.map(({ label, value }) => (
                  <option key={label + value} value={value}>
                    {label}
                  </option>
                ))}
              </select>
            ) : null}

            {filterVariant === "checkBox" ? (
              <div className="filter-checkbox-input">
                <label>{placeholder}</label>
                <input
                  type="checkbox"
                  value={String(columnFilterValue?.value ?? true)}
                  onChange={(event_) => {
                    handleChange({
                      value: event_.target.checked,
                      filterFn: filterFn || "equals",
                    });
                  }}
                />
              </div>
            ) : null}
          </div>
        </OutsideClickHandler>
      ) : null}
    </div>
  );
}

export default Filter;
