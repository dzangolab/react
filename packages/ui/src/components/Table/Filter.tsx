import React, { useContext, useState } from "react";
import { DebounceInput } from "react-debounce-input";
import OutsideClickHandler from "react-outside-click-handler";

import { TableContext } from "./TableProvider";

import type { FilterProperties } from "./types";

function Filter({ column, table }: FilterProperties) {
  const { filterDebounceTime } = useContext(TableContext);

  const [expanded, setExpanded] = useState(false);

  const firstValue = table
    .getPreFilteredRowModel()
    .flatRows[0]?.getValue(column.id);

  const columnFilterValue = column.getFilterValue();

  const toggleExpand = () => {
    expanded ? setExpanded(false) : setExpanded(true);
  };

  return (
    <div
      className="filter-wrapper"
      onClick={(event_) => event_.stopPropagation()}
    >
      {<button onClick={toggleExpand}>{expanded ? "ğŸ™‰" : "ğŸ™ˆ"}</button>}
      {expanded ? (
        <OutsideClickHandler
          onOutsideClick={() => {
            setExpanded(false);
          }}
        >
          <div className="filter-form-wrapper">
            {typeof firstValue === "number" ? (
              <div className="number-filter-form">
                <DebounceInput
                  type="number"
                  value={(columnFilterValue as [number, number])?.[0] ?? ""}
                  debounceTimeout={filterDebounceTime}
                  onChange={(event_) => {
                    column.setFilterValue((old: [number, number]) => [
                      event_.target.value,
                      old?.[1],
                    ]);
                  }}
                  placeholder={`Min`}
                  className=""
                  size={2}
                />

                <DebounceInput
                  type="number"
                  value={(columnFilterValue as [number, number])?.[1] ?? ""}
                  debounceTimeout={filterDebounceTime}
                  onChange={(event_) => {
                    column.setFilterValue((old: [number, number]) => [
                      event_.target.value,
                      old?.[0],
                    ]);
                  }}
                  placeholder={`Max`}
                  className=""
                  size={2}
                />
              </div>
            ) : (
              <DebounceInput
                type="text"
                value={(columnFilterValue ?? "") as string}
                debounceTimeout={filterDebounceTime}
                onChange={(event_) => {
                  column.setFilterValue(event_.target.value);
                }}
                placeholder={`Search...`}
                className=""
                size={10}
              />
            )}
          </div>
        </OutsideClickHandler>
      ) : null}
    </div>
  );
}

export default Filter;
