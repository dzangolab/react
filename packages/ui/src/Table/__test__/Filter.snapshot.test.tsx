import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test, vi } from "vitest";

import { columns, data, fetcher } from "./TestTableData";
import Filter from "../common/Filter";
import { Table } from "../index";

const mockFunction = vi.fn();

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Table
      columns={columns}
      data={data}
      fetcher={fetcher}
      totalItems={data.length}
    >
      {/* <Filter column={} /> */}
    </Table>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
