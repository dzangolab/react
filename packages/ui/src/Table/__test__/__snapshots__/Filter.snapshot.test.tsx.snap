// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div
  className="filter-wrapper"
  onClick={[Function]}
>
  <button
    onClick={[Function]}
  >
    <img />
  </button>
</div>
`;
